book-content/chapters/14-configuring-typescript.md

```diff
@@ -0,0 +1,1024 @@
+    "composite": true,
+    "declarationMap": true,
+
+    /* If NOT transpiling with tsc: */
+    "module": "Preserve",
+    "noEmit": true,
+
+    /* If your code runs in the DOM: */
+    "lib": ["es2022", "dom", "dom.iterable"],
+
+    /* If your code doesn't run in the DOM: */
+    "lib": ["es2022"]
+  }
+}
+```
+
+Now that we understand the lay of the land, let's take a look at each of these options in more detail.
+
+## Base Options
+
+### `target`
+
+The `target` option specifies the ECMAScript version that TypeScript should target when generating JavaScript code.
+
+For example, setting `target` to `ES5` will attempt to transform your code to be compatible with ECMAScript 5.
+
+Language features like optional chaining and nullish coalescing, which were introduced later than ES5, are still available:
+
+```tsx
+// Optional chaining
+const search = input?.search;
+
+// Nullish coalescing
+const defaultedSearch = search ?? "Hello";
+```
+
+But when they are turned into JavaScript, they'll be transformed into code that works in ES5 environments:
+
+```javascript
+// Optional chaining
+var search = input === null || input === void 0 ? void 0 : input.search;
+// Nullish coalescing
+var defaultedSearch = search !== null && search !== void 0 ? search : "Hello";
+```
+
+#### `target` Doesn't Polyfill
+
+While `target` can transpile newer syntaxes into older environments, it won't do the same with API's that don't exist in the target environment.
+
+For example, if you're targeting a version of JavaScript that doesn't support `.replaceAll` on strings, TypeScript won't polyfill it for you:
+
+```tsx
+const str = "Hello, world!";
+
+str.replaceAll("Hello,", "Goodbye, cruel");
+```
+
+This code will error in your target environment, because `target` won't transform it for you. If you need to support older environments, you'll need to find your own polyfills. You configure the environment your code executes in with `lib`, as we saw in a previous chapter.
+
+If you're not sure what to specify for `target`, keep it up to date with the version you have specified in `lib`.
+
+### `esModuleInterop`
+
+`esModuleInterop` is an old flag, released in 2018. It helps with interoperability between CommonJS and ES modules. At the time, TypeScript had deviated slightly from commonly-used tools like Babel in how it handled wildcard imports and default exports. `esModuleInterop` brought TypeScript in line with these tools.
+
+You can read the [release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#support-for-import-d-from-cjs-from-commonjs-modules-with---esmoduleinterop) for more details. Suffice to say, when you're building an application, `esModuleInterop` should always be turned on. There's even a proposal to make it the default in TypeScript 6.0.
+
+### `isolatedModules`
+
+`isolatedModules` prevents some TypeScript language features that single-file transpilers can't handle.
```

Comment:
More importantly, it's a mode that bans ambiguous situations like:
```ts
export { Foo } from './foo'
export const bar = 42;
```

What should be emitted here as JS? Should it be:
```ts
export { Foo } from './foo.js'
export const bar = 42
```
or something closer to this?
```ts
import './foo.js'
export const bar = 42
```

The problem is that based on this single file we can't possibly know if `Foo` is a value or a type and if it can be referenced in the emitted JS.